Feat/#409-K: 선택된 회의록 id URL에 반영

se030 opened the pull request:
## 🤠 개요

<!-- 

- 이슈번호
- 한줄 설명
 
-->

- Closes #409 

- 선택 회의록 id를 URL에 추가하고, 전역 상태를 사용해 불필요한 EventEmitter 사용을 제거했어요.

## 💫 설명

<!-- 

- 현재 Pr 설명 

-->

- pathname 정규식이 수정됐어요. 정규식 이해를 돕기 위해 주석을 추가했습니다.

  ```ts
  // App.tsx
    const validPathPattern = /^\/workspace(\/\d+(\/.+)?)?$/; // /workspace(/숫자(/아무거나)) 와 처음부터 끝까지 일치하는 패턴
    if (user && !validPathPattern.test(pathname)) {
      navigate('/workspace');
    }

  // ...
          <Routes>
            <Route path="/" element={<LoginPage />} />
            <Route path="/oauth" element={<OAuthPage />} />
            <Route path="/workspace/*" element={<WorkspacePage />} />
  ```

- Workspace 컴포넌트에서 선택 회의록 id를 알 수 있고 URL에 따라서 동작합니다.

  ```ts
  // Workspace.tsx
  const params = useParams();
  const momId = params['*'];

  // ...

  useEffect(() => {
    if (!workspace) return;
    const { moms } = workspace;

    if (!momId && moms.length) {
      navigate(moms[0]._id); // 선택 회의록이 없는 경우 첫번째 회의록으로 navigate
    }
  }, [workspace, momId]);
  ```

- MomList 아이템들의 onSelect 핸들러는 선택 회의록 정보를 리셋하고 선택된 id로 navigate합니다.

  ```ts
  // MomList.tsx
    const onSelect = (id: string) => {
      setSelectedMom(null);
      navigate(id);
    };
  ```

- Workspace 쪽에서 URL 정보를 사용해 `MOM_EVENT.SELECT`를 emit하는 흐름이에요. 동일 이벤트에 의해 발생하는 동작 흐름이 비슷한 위치에 있어야 이해하기 쉬울 것 같아 관련 로직은 모두 Workspace 컴포넌트로 옮겼습니다.

  ```ts
  // 선택 회의록이 변경되면 클라이언트 상태를 변경하기 전에 회의록 소켓 서버에 알려줘야 함
  useEffect(() => {
    if (momId && momSocket) {
      const message: MomMessage.Select = { id: momId };
      momSocket.emit(MOM_EVENT.SELECT, message);
    }
  }, [momId, momSocket]);
  ```

## 🌜 고민거리

<!-- 

### Q. 고민1
뭐시기 뭐시기 

-->

- 작업하면서 상태관리 라이브러리 사용이 결정돼서 Context 관련 리팩토링은 진행하지 않을게요.

  - Closes #410

- 마이그레이션 작업에서 앱 전체적으로 상태에 대한 관심사가 제대로 분리되어 있는지, 비동기로 업데이트되는 상태들에 대해서 로딩이나 Skeleton이 제대로 적용되고 있는지 확인이 필요할 것 같아요.

  - 아래처럼 Skeleton 대신 빈 영역이 렌더링되고 있어 해결 방법을 고민하고 있어요. (상태 관련 리팩토링이 필요해요.)
  
    <img width="500" alt="image" src="https://github.com/boostcampwm-2022/web27-Wabinar/assets/63814960/a025ab3c-9cce-49ad-892b-e0441ca3047d">

## 📷 스크린샷 (Optional)

se030 committed:
feat: 선택된 회의록 id url params에 추가

se030 committed:
feat: URL 통한 회의록 선택 기능 추가

se030 committed:
refactor: navigation, 회의록 상태 및 이벤트 로직 위치 이동

se030 commented on client/src/components/Sidebar/MomList.tsx:
```diff
@@ -2,57 +2,43 @@ import { RiFileAddLine } from '@react-icons/all-files/ri/RiFileAddLine';
 
 interface MomListProps {
   moms: TMom[];
-  selectedMom: TMom | null;
```
props로 내릴 필요가 없는 것 같아 수정했는데, 다시보니 Context 리렌더링 방지를 위해서 props + memo를 사용한 부분이었어요. 근데 이제보니 리렌더링을 막기보다는 MomList에서도 selectedMom 상태 반영해서 선택 회의록 스타일링 해주면 좋을 것 같은데 어떨까요?

dohun31 replied:
헉 좋아요
사실 저거 리렌더링은 성능에 부하를 줄 정도도 아니더라구요... ㅋㅋㄱㅋㅋ
왜 저렇게 집착했는지😹

선택한 회의록 스타일링 해주는거 사용자 입장에서 너무 좋을것같아요!!!!!

se030 replied:
그것은 메모 박사가 되기 위한 과정이죠 ..🤓 css 소소하게 고쳐볼게요!

(done) [feat: 선택 회의록 정보 MomList 스타일에 반영](https://github.com/boostcampwm-2022/web27-Wabinar/pull/411/commits/ea184e717e2f29ab011c35535867e1669588532d)

se030 committed:
chore: recoil 설치

se030 committed:
feat: Mom 로딩 상태 및 DefaultMom 렌더링 로직의 workspace.moms에 recoil 적용

se030 committed:
Merge branch 'dev' into feat/#409-K

se030 committed:
docs: 회의록 선택에 따른 URL 변경 및 SELECT 이벤트 흐름 주석 추가

se030 committed:
refactor: 회의록 선택 로직 MomList로 이동

se030 committed:
del: 중복 로직 제거

dohun31 commented on client/src/components/Mom/index.tsx:
```diff
@@ -3,17 +3,20 @@ import * as MomMessage from '@wabinar/api-types/mom';
 import style from './style.module.scss';
 
 function Mom() {
+  const [workspace] = useRecoilState(workspaceState);
```
```suggestion
  const workspace = useRecoilValue(workspaceState);
```

se030 replied:
아래 코멘트로 반영할게요!

dohun31 commented on client/src/components/Mom/index.tsx:
```diff
@@ -230,32 +231,19 @@ function Mom() {
 
-  if (!isLoaded) {
+  if (!workspace) {
     return <div className={style['mom-container']}></div>;
   }
 
-  if (isMomsEmpty && !selectedMom) {
+  if (!workspace.moms.length) {
```
갑자기 든 생각인데!!
`workspace` 상태를 `Mom`이 알아야할까요?
Workspace 컴포넌트에서 `Suspense`처리나 `<DefaultMom />` 렌더링 해주는건 어때요??

저 코드 짤땐 생각이 안났는데 뭔가 깔끔해지니까 보이네요?! ㅋㅋㅋㅋㅋ

se030 replied:
천재 아니에요? 어차피 컴포넌트 트리에서 Workspace가 Mom 상위에 있으니까 이게 맞는 것 같아요 ...

dohun31 replied:
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ진짜 오랜만에 듣는 밈

se030 replied:
[refactor: 회의록 없는 경우의 분기 Mom에서 Workspace 컴포넌트로 이동](https://github.com/boostcampwm-2022/web27-Wabinar/pull/411/commits/07c6c331265a8af578ad521d46f19af3b10abdf0)

dohun31 reviewed (commented):
진짜 상태들이 다 여기저기 흩어져있어서 애먹었네요
관심사 분리 시급해요,,,😂

dohun31 commented:
각 폴더들에 `index.ts` 만들어서 import문 정리하는 의견 제시합니다..

이유는 너무 `import ... from ...`이 많아서 import순서가 섞여있으면 뭔가 보기 힘들어요

```tsx
import { Workspace, Mom, MomList } from "@src/components"
import { 
  SelectedMomContext, 
  WorkspaceContext,
 } from "@src/hooks/context"
```

se030 commented:
> 각 폴더들에 `index.ts` 만들어서 import문 정리하는 의견 제시합니다..
> 
> 이유는 너무 `import ... from ...`이 많아서 import순서가 섞여있으면 뭔가 보기 힘들어요
> 
> ```tsx
> import { Workspace, Mom, MomList } from "@src/components"
> import { 
>   SelectedMomContext, 
>   WorkspaceContext,
>  } from "@src/hooks/context"
> ```

배럴 파일로 만들면 원본 파일을 찾아갈 때 두 depth를 들어가야 하는데 어떤게 더 불편할까요?
이제보니 자동완성 경로도 `src/components`, `comopnents/` 이런식으로 섞여있네요. 여러가지 대공사가 필요해요.

dohun31 commented:
> > 각 폴더들에 `index.ts` 만들어서 import문 정리하는 의견 제시합니다..
> > 이유는 너무 `import ... from ...`이 많아서 import순서가 섞여있으면 뭔가 보기 힘들어요
> > ```tsx
> > import { Workspace, Mom, MomList } from "@src/components"
> > import { 
> >   SelectedMomContext, 
> >   WorkspaceContext,
> >  } from "@src/hooks/context"
> > ```
> 
> 배럴 파일로 만들면 원본 파일을 찾아갈 때 두 depth를 들어가야 하는데 어떤게 더 불편할까요? 이제보니 자동완성 경로도 `src/components`, `comopnents/` 이런식으로 섞여있네요. 여러가지 대공사가 필요해요.

뭐가됐든 먼저 하나로 통일 시켜야겠네요

dohun31 commented:
아 정규식도 상수로 빼서 관리하는건 어때요?

se030 committed:
refactor: 회의록 없는 경우의 분기 Mom에서 Workspace 컴포넌트로 이동

se030 committed:
refactor: Sidebar의 workspace recoil atom으로 변경

se030 committed:
feat: 선택 회의록 정보 MomList 스타일에 반영

se030 committed:
feat: MOM_EVENT.SELECT emit 위치 Workspace로 이동

- 초기 접속한 URL에 momId가 지정되어 있을 경우도 커버하기 위해 URL 변경에 소켓 이벤트가 발생하도록 변경
- MomList는 navigate 역할만 수행

se030 committed:
feat: 워크스페이스 이동 로직 개선

- 이동할 워크스페이스 정보가 로드되기 전 렌더링 로직을 위해 workspace atom null로 리셋
- 이전과 값이 동일한 경우에는 변동이 없도록 수정

se030 commented:
> 아 정규식도 상수로 빼서 관리하는건 어때요?

정규식 특성 상 가독성이 떨어지긴 하는데, 어떻게 명확한 네이밍을 할지도 고민이고 지금 사용되는 정규식 패턴이 서로 달라서 상수화가 의미있을지 잘 모르겠어요. 의미를 쉽게 파악할 수 있게 주석을 한번 추가해볼게요!

정규식을 읽기 쉽게 쓸 수 있게 해주는 아래와 같은 라이브러리들도 있는데 두 라인을 위해서 도입하기에는 과한 느낌이에요. 언젠가 유용하게 쓸 수 있을지 모르니 남겨둘게요 😗

- https://github.com/VerbalExpressions/JSVerbalExpressions

- https://github.com/danielroe/magic-regexp

se030 committed:
refactor: MOM_EVENT.SELECT 이벤트 관련 코드 Workspace로 이동

se030 committed:
docs: 정규식 주석 추가

se030 commented:
PR Details 수정했어요 리뷰하실 때 참고해주세요 👍🏻

se030 committed:
refactor: momId 추출에 정규식 대신 react-router useParams 사용

dohun31 commented on client/src/components/Sidebar/index.tsx:
```diff
@@ -1,29 +1,27 @@
-          selectedMom={selectedMom}
-          setSelectedMom={setSelectedMom}
-        />
+        <MemberList members={members} />
+        <MomList moms={moms} />
```
우와 너무 깔끔해졌어요 ㅋㅋㅋㅋㅋ

dohun31 commented on client/src/components/Workspace/index.tsx:
```diff
@@ -28,7 +35,9 @@ function Workspace() {
       setWorkspace(workspaceInfo);
 
-      if (!workspaceInfo.moms[0]) setSelectedMom(null);
+      if (!workspaceInfo.moms.length) {
```
오 👍

dohun31 commented on client/src/components/Workspace/index.tsx:
```diff
@@ -72,8 +109,8 @@ function Workspace() {
-            <Sidebar workspace={workspace} />
-            <Mom />
+            <Sidebar />
+            {workspace.moms.length ? <Mom /> : <DefaultMom />}
```
여기가 맞았어요.... ㅋㅋㅋㅋ

se030 replied:
맞아요 그때는 왜 이게 안보였는지 ......ㅎㅎ

dohun31 commented on client/src/components/WorkspaceThumbnailList/index.tsx:
```diff
@@ -9,9 +11,14 @@ interface WorkspaceThumbnailListProps {
 function WorkspaceThumbnailList({ workspaces }: WorkspaceThumbnailListProps) {
+  const { id: currentId } = useParams();
   const navigate = useNavigate();
+  const [, setWorkspace] = useRecoilState(workspaceState);
```
```suggestion
  const setWorkspace = useSetRecoilState(workspaceState);
```
- [useSetRecoilState](https://recoiljs.org/docs/api-reference/core/useSetRecoilState/)

아니면 나중에 workspace 사용할 수 도 있어서 그래요??

se030 replied:
헉 setState만 가져오는 API가 있는지 몰랐어요. 하지만 #412 에 따라 기각 ...

dohun31 commented on client/src/components/WorkspaceThumbnailList/index.tsx:
```diff
@@ -9,9 +11,14 @@ interface WorkspaceThumbnailListProps {
   const onClick = (targetId: number) => {
+    if (Number(currentId) === targetId) return;
+
+    setWorkspace(null);
```
🥇

dohun31 replied:
workspace 없다고 알려주고 
-> Workspace 컴포넌트가 mount돼면 (id가 바뀌면)
-> loadWorkspaceInfo 함수 실행해서 workspace 바꾸기

흐름이 이렇게 되는거 맞나요?

dohun31 reviewed (approved):
너무 깔끔해졌어요
이제야 각자의 관심사를 찾아간 느낌이에요 ㅋㅋㅋㅋ
세영님 대단해요

se030 committed:
chore: react-query 설치

Refactor: 상태 관리에 react-query 적용 #412

se030 committed:
feat: workspace 상태에 useQuery 사용

recoil 관련 코드 제거 및 react-query로 대체
